shopfloor_reception/services/reception.py

```diff
@@ -302,91 +322,165 @@ def _scan_document__by_picking(self, barcode):
+        # scanned a wrong product. Warn him about this.
+        if origin_moves and not origin_moves_for_product:
+            message = self.msg_store.product_not_found_in_picking()
+            return self._response_for_select_move(picking, message=message)
+        # If we have an origin move, create the return move, and go to next screen
+        if origin_moves_for_product:
+            return_move = self._scan_line__create_return_move(
+                picking, origin_moves_for_product
             )
-            message = self._check_move_available(move, message_code="packaging")
-            if message:
-                return self._response_for_select_move(
-                    picking,
-                    message=message,
-                )
-            return self._select_line_from_packaging(picking, move, packaging)
+            if not return_move:
+                # It means that among all origin moves, none has been found with
+                # max qty to return being positive.
+                # Which means all lines have already been returned.
+                message = self.msg_store.move_already_returned()
+                return self._response_for_select_move(picking, message=message)
+            picking.action_confirm()
+            picking.action_assign()
+            # return_move._action_confirm()
+            return self._scan_line__find_or_create_line(picking, return_move)
+
+    def _scan_line__by_product(self, picking, product):
+        if self.work.menu.allow_return:
+            response = self._scan_line__by_product__return(picking, product)
+            if response:
+                return response
+        # Otherwise, the picking isn't a return, and should be a regular reception
+        move = picking.move_lines.filtered(lambda m: m.product_id == product)
+        message = self._check_move_available(move, "product")
+        if message:
+            return self._response_for_select_move(
+                picking,
+                message=message,
+            )
+        return self._scan_line__find_or_create_line(picking, move)
 
-    def _scan_line__by_lot(self, picking, barcode):
-        line = picking.move_line_ids.filtered(
-            lambda l: barcode == l.lot_id.name or barcode == l.lot_name
-        )
-        move = line.move_id
+    def _scan_line__by_packaging__return(self, picking, packaging):
         search = self._actions_for("search")
-        lot = search.lot_from_scan(barcode)
-        if not move:
-            line = picking.move_line_ids.filtered(
-                lambda l: not l.lot_id
-                and not l.lot_name
-                and l.product_id == lot.product_id
+        origin_moves = search.origin_move_from_scan(picking.origin)
+        origin_moves.filtered(lambda m: m.picking_id.picking_type_code == "outgoing")
+        origin_moves_for_packaging = origin_moves.filtered(
+            lambda m: packaging in m.product_id.packaging_ids
+        )
```

Comment:
inefficient, give a domain to the search

Reply:
done